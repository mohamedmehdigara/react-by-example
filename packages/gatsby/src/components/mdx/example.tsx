import React, { FC, ReactNode, useCallback, useState } from 'react'
import styled from 'styled-components'
import { CodeSandboxIcon, ResetIcon } from '../icons'

const Render = styled.div`
  display: flex;
  padding: 1em;
  border: 1px solid var(--app-foreground);
  border-radius: var(--app-border-radius);
  align-items: center;
  justify-content: center;
`
const FlexRight = styled.div`
  display: flex;
  justify-content: flex-end;
`

const CodeSandboxLink = styled.a`
  display: flex;
`

const ResetButton = styled.button`
  display: flex;
  margin: 0;
  padding: 0;
  color: var(--app-foreground);
  background: none;
  border: none;
  cursor: pointer;

  :focus,
  :active {
    color: var(--app-cyan);
    outline: 2px solid var(--app-cyan);
  }
`

export interface ExampleProps {
  code?: ReactNode
  render?: ReactNode
  link?: string
}

export const Example: FC<ExampleProps> = ({ code, render, link }) => {
  const [resetKey, setResetKey] = useState(Date.now())
  const triggerReset = useCallback(() => {
    setResetKey(Date.now())
  }, [])
  return (
    <>
      {code}
      <Render key={resetKey}>
        <div>{render}</div>
      </Render>
      {link && (
        <FlexRight>
          <ResetButton
            type="button"
            title="reset example"
            onClick={triggerReset}
          >
            <ResetIcon />
          </ResetButton>
          <CodeSandboxLink
            target="_blank"
            rel="noreferrer"
            href={link}
            title="edit on CodeSandbox"
          >
            <CodeSandboxIcon />
          </CodeSandboxLink>
        </FlexRight>
      )}
    </>
  )
}
