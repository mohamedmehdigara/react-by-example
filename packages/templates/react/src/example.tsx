import React, { FC, ReactElement } from 'react'

export const Simple: FC = () => <h1>Hello world!</h1>

export default (): ReactElement => <Simple />
