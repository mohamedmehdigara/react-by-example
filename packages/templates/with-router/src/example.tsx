import React, { FC } from 'react'

export const Simple: FC = () => <h1>Hello world!</h1>

export default <Simple />
