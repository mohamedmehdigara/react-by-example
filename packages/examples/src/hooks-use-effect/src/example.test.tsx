import {
  act,
  cleanup,
  fireEvent,
  render,
  RenderResult,
} from '@testing-library/react'
import React from 'react'
import { LocaleClock, ToggleLocale } from './example'

let time: string
const useMockTime = () => {
  const originalDate = Date
  const mockDate = class extends Date {
    constructor() {
      super(`2020-04-01 ${time}`)
    }
  }

  beforeEach(() => {
    time = '12:34:00'
    Object.assign(window, { Date: mockDate })
  })

  afterEach(() => {
    Object.assign(window, { Date: originalDate })
  })
}

describe('LocaleClock', () => {
  let component: RenderResult
  let setIntervalSpy: jest.SpyInstance
  let clearIntervalSpy: jest.SpyInstance

  useMockTime()

  beforeEach(() => {
    jest.useFakeTimers()
    setIntervalSpy = jest.spyOn(window, 'setInterval')
    clearIntervalSpy = jest.spyOn(window, 'clearInterval')
  })
  afterEach(() => {
    jest.restoreAllMocks()
    jest.useRealTimers()
  })

  describe('default locale', () => {
    beforeEach(() => {
      component = render(<LocaleClock />)
    })

    afterEach(() => {
      cleanup()
    })

    it('should snapshot', () => {
      expect(component.container).toMatchSnapshot()
    })

    it('should show time', () => {
      expect(component.container).toHaveTextContent('12:34:00')
    })

    it('should set an interval', () => {
      expect(setIntervalSpy).toHaveBeenCalledWith(expect.any(Function), 500)
      expect(setIntervalSpy).toHaveBeenCalledTimes(1)
    })

    describe('when time has passed', () => {
      beforeEach(() => {
        time = '12:34:01'
        act(() => {
          jest.advanceTimersByTime(1000)
        })
      })

      it('should show time', () => {
        expect(component.container).toHaveTextContent('12:34:01')
      })

      describe('when time has passed', () => {
        beforeEach(() => {
          time = '12:34:02'
          act(() => {
            jest.advanceTimersByTime(1000)
          })
        })

        it('should show time', () => {
          expect(component.container).toHaveTextContent('12:34:02')
        })

        describe('when time has passed', () => {
          beforeEach(() => {
            time = '12:34:03'
            act(() => {
              jest.advanceTimersByTime(1000)
            })
          })

          it('should show time', () => {
            expect(component.container).toHaveTextContent('12:34:03')
          })
        })
      })
    })

    it('should not have cleared the interval', () => {
      expect(clearIntervalSpy).not.toHaveBeenCalled()
    })

    describe('when unmounted', () => {
      beforeEach(() => {
        cleanup()
      })

      it('should clear the interval', () => {
        expect(clearIntervalSpy).toHaveBeenCalledWith(expect.any(Number))
        expect(clearIntervalSpy).toHaveBeenCalledTimes(1)
      })
    })
  })

  describe('specific locale', () => {
    beforeEach(() => {
      component = render(<LocaleClock locale="en-US" />)
    })

    afterEach(() => {
      cleanup()
    })

    it('should show time', () => {
      expect(component.container).toHaveTextContent('12:34:00 PM')
    })
  })
})

describe('ToggleLocale', () => {
  let component: RenderResult
  let toggle: HTMLElement

  beforeEach(() => {
    jest.useFakeTimers()
    jest
      .spyOn(Date.prototype, 'toLocaleTimeString')
      .mockImplementation(
        (locale) => `current time in "${String(locale)}" locale`,
      )
    component = render(<ToggleLocale />)
    toggle = component.getByText('toggle locale')
  })

  afterEach(() => {
    cleanup()
    jest.restoreAllMocks()
    jest.useRealTimers()
  })

  it('should snapshot', () => {
    expect(component.container).toMatchSnapshot()
  })

  it('should show time', () => {
    expect(component.container).toHaveTextContent(
      'current time in "en-GB" locale',
    )
  })

  describe('when toggling', () => {
    beforeEach(() => {
      fireEvent.click(toggle)
    })

    it('should show time', () => {
      expect(component.container).toHaveTextContent(
        'current time in "th-TH-u-nu-thai" locale',
      )
    })
  })
})
