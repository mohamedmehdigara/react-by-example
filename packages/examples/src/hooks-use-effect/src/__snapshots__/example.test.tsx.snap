// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LocaleClock default locale should snapshot 1`] = `
<div>
  <p>
    12:34:00 PM
  </p>
</div>
`;

exports[`ToggleLocale should snapshot 1`] = `
<div>
  <button
    type="button"
  >
    toggle locale
  </button>
  <p>
    current time in "en-GB" locale
  </p>
</div>
`;
